src/frontend/rect-controller.tsx

```diff
@@ -72,3 +62,13 @@ export function RectController({
     </DraggableCore>
   );
 }
+
+export const RectController = React.memo(
+  RectControllerInternal,
+  (prev, next) => {
+    return (
+      isEqual(prev.shape, next.shape) &&
+      shallowequal({ ...prev, shape: undefined }, { ...next, shape: undefined })
```

Comment:
Instead of introducing this, we could handle validation differently: `parse()` the incoming object to validate it, but don't return the object, return the original one. That way memo keeps working by identity.